### PR TITLE
Long press

### PR DESCRIPTION
**Pull request by nperozzi**
Hi,
I implemented a long-press button.
This is the goal: A press of over 3 seconds qualifies as a long press. Since it is difficult to tell when you reached 3 secs, I want the user to keep the button pressed and when the 3 secs is reached, the system performs the action. Once the user sees that the action has been performed (ie, receives the feedback that the long press had its effect), the user will release the button. 
I believe this is how long press works on most devices. Looking forward to hearing what you think :)

I have also implemented the double button press.

PS: I am doing a PS mostly to share this peace of code and discuss it, not cause I believe it is ready to merge.

### CODE REVIEW

File: lib/atmega328/rotary-encoder.cpp [hunk 1]
```diff
@@ -59,14 +63,63 @@ ISR(INT0_vect)
     }
 }
 
+#define LONG_PRESS_DURATION 3000
+#define DOUBLE_PRESS_DURATION 500
+volatile bool button_pressed = false;
+volatile unsigned long button_press_start = 0;
+volatile uint8_t button_press_count = 0;
```
Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:70]: The state is determined by the collection of these variables' values. They should thus be grouped into a single object / `struct`.


```c
struct button {
    bool pressed_down;
    unsigned long press_start_time_ms;
    uint8_t press_count;
};
```

File: lib/atmega328/rotary-encoder.cpp [hunk 2]
```diff
@@ -59,14 +63,63 @@ ISR(INT0_vect)
     }
 }
 
+#define LONG_PRESS_DURATION 3000
+#define DOUBLE_PRESS_DURATION 500
```
Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:67]: Prefer typed values before defines. That way you also get a reasonable compiler warning / error if you're doing something strange.

```suggestion
const uint16_t long_press_duration_ms = 3000;
const uint16_t double_press_duration_ms = 500;
```

Pre-processor statements can potentially give more headaches, and should not be resorted to unless necessary.

Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:67]: I think 3000 ms is a bit too long for long press. Perhaps around 1 second should be enough?

File: lib/atmega328/rotary-encoder.cpp [hunk 3]
```diff
@@ -59,14 +63,63 @@ ISR(INT0_vect)
...
+volatile bool button_pressed = false;
+volatile unsigned long button_press_start = 0;
```
Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:69]: Use explicit type `uint16_t` instead of relying on the compiler's `unsigned long` type

File: lib/atmega328/rotary-encoder.cpp [hunk 4]
```diff
@@ -59,14 +63,63 @@ ISR(INT0_vect)
...
 ISR(PCINT2_vect)
 {
     if (should_retrigger_after_sw_debounce(&last_trigger_PCINT0))
     {
-        if (bit_is_set(PIND, SW_PIN))
+        if (bit_is_clear(PIND, SW_PIN))
         {
-            button_press();
+            if (!button_pressed)
+            {
```
Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:79]: Instead of having nested `if`-statements with lots of indentation (harder to read), you should merge them when they are logically grouped

```c
if (bit_is_clear(PIND, SW_PIN) && !button_pressed) {
```

Comment (emanuelen5) [reply lib/atmega328/rotary-encoder.cpp:79]: Also, make sure to create variables that explains what the if-statements check

```c
const bool button_went_from_released_to_pressed = bit_is_clear(PIND, SW_PIN) && !button_pressed;
const bool button_went_from_pressed_to_released = !bit_is_clear(PIND, SW_PIN) && button_pressed;

if (button_went_from_released_to_pressed) {
    button_pressed = true;
    button_press_start = millis();
    button_press_count++;
} else if (button_went_from_pressed_to_released) {
    button_pressed = false;
}
```

File: lib/atmega328/rotary-encoder.cpp [hunk 5]
```diff
@@ -59,14 +63,63 @@ ISR(INT0_vect)
...
+    return (millis() - button_press_start);
+}
+
+void service_button_press()
+{
+    if(!button_pressed && button_press_count == 1 && button_press_timer() > DOUBLE_PRESS_DURATION)
+    {
+        single_button_press();
+        reset_button_press();
+    }
```
Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:112]: I think you should either create helper (object) methods for these `if`-statements or helper variables (example below):

```c
const bool button_was_single_pressed = !button_pressed && button_press_count == 1 && button_press_timer() > DOUBLE_PRESS_DURATION;

if (button_was_single_pressed) {
    single_button_press();
    reset_button_press();
}

// etc.
```

File: lib/atmega328/rotary-encoder.cpp [hunk 6]
```diff
@@ -59,14 +63,63 @@ ISR(INT0_vect)
...
+void reset_button_press()
+{
+    button_press_count = 0;
+    button_press_start = 0;
+}
+
+unsigned long button_press_timer()
+{
+    return (millis() - button_press_start);
+}
```
Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:104]: Perhaps the name `time_since_initial_button_press()` makes more sense.

Also, prefer explicit types, i.e. `uint16_t`.

File: lib/atmega328/rotary-encoder.cpp [hunk 7]
```diff
@@ -59,14 +63,63 @@ ISR(INT0_vect)
...
+    {
+        double_button_press();
+        reset_button_press();
+    }
+
+    if(button_pressed && button_press_timer() >= LONG_PRESS_DURATION)
+    {
+        long_button_press();
+        reset_button_press();
     }
```
Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:124]: I think the logic is flawed, and the function has a few bugs. The point in time when the function is called cannot be assumed! :)

1. if the button was pressed twice, but the service routine wasn't pressed until it had reached 500 ms, then nothing will happen (but it should have been a double-press)
1. if you manage to press the button more than 2 times before the service routine is called (or hold it down for 500 ms and then let it go), then it will also get stuck and might falsely give you a _long press_ (even though it was a double-press) when you finally press it down one more time

Comment (emanuelen5) [reply lib/atmega328/rotary-encoder.cpp:124]: It would make sense to make this function unit-test:able so we can figure out those edge-cases.

**Review by emanuelen5 — COMMENTED** [review]
Nice solution. I think we mostly need to fix the single/double press logic, and perhaps polish it a bit. :)

File: lib/atmega328/rotary-encoder.cpp (hunk 1 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp:70]: That is good! I just added the button struct

File: lib/atmega328/rotary-encoder.cpp (hunk 2 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp:67]: I see. I will start using const then 😅. Change implemented.

File: lib/atmega328/rotary-encoder.cpp (hunk 3 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp:69]: True. Unsigned long is compiler-dependent.

File: lib/atmega328/rotary-encoder.cpp (hunk 2 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp:67]: Let´s compromise in 2000ms and we see how it feels. If, later on, it feels still too long, we can modify it. 😉

File: lib/atmega328/rotary-encoder.cpp (hunk 4 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp]: That would make it so much easier to read. Done

File: lib/atmega328/rotary-encoder.cpp (hunk 5 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp]: haha I did it while fixing the previous comment 😅

File: lib/atmega328/rotary-encoder.cpp (hunk 7 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp]: This one left me thinking... I need to sit and put some thought into this. 🤔

File: lib/atmega328/rotary-encoder.cpp [hunk 8]
```diff
@@ -59,14 +66,63 @@ ISR(INT0_vect)
...
+        const bool button_goes_from_pressed_to_released = !bit_is_clear(PIND, SW_PIN) && button.pressed_down;
+        if (button_goes_from_released_to_pressed)
         {
-            button_press();
+            button.pressed_down = true;
+            button.press_start_time_ms = millis();
+            button.press_count++;
+        } else if (button_goes_from_pressed_to_released) {
+            button.pressed_down = false;
         }
```
Comment (emanuelen5) [inline lib/atmega328/rotary-encoder.cpp:86]: It becomes evident to me now that the button should be a separate class/struct. That would make unit-testing easy. However, we can leave that to future PR.

**Review by emanuelen5 — DISMISSED** [review]
Nice job! The change history was also really easy to follow for the last few commits. 👍

File: lib/atmega328/rotary-encoder.cpp (hunk 6 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp]: Done

File: lib/atmega328/rotary-encoder.cpp [hunk 9]
```diff
@@ -59,14 +66,63 @@ ISR(INT0_vect)
     }
 }
 
-unsigned long last_trigger_PCINT0 = 0;
+const uint16_t long_press_duration_ms = 2000;
+const uint16_t double_press_duration_ms = 500;
+uint16_t last_trigger_PCINT0 = 0;
```
[the remark this answers is not in the document]
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp:71]: true ture. Change made 😄

File: lib/atmega328/rotary-encoder.cpp (hunk 8 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp:86]: The button is already a struct with its own member variable. Do you mean that it should be its own library?  Maybe, I need more explanation here. 🐌

Comment (emanuelen5) [reply lib/atmega328/rotary-encoder.cpp:86]: Yeah, it would have to be put into its own file. A file that just contains the functionality for pressing the button (not anything relating to the rotation). The button can be described as a state machine, and I think that would make it a lot easier to avoid the sneaky edge conditions :)

But either way, I think we should do this in a follow-up PR. This is definitely good enough for now!

I started a sketch on the state diagram, but it became a mess. I'll have to get back to it. Leaving it here for now:

```mermaid
stateDiagram-v2
    state if_state1 <<choice>>
    state if_state2 <<choice>>

    Init --> EvaluateSinglePress: PRESS
    EvaluateSinglePress --> SinglePress: [time_since_init < 500 ms] DEPRESS
    EvaluateSinglePress --> LongPress: [time of state > 2000 ms]

    state EvaluateSinglePress {
        [*] --> S1_Wait: [time_since_init <= 500 ms] DEPRESS
    }
    EvaluateSinglePress --> if_state1
    if_state1 --> SinglePress: [time_since_init > 500ms]
    if_state1 --> EvaluateDoublePress: [time_since_init <= 500ms]

    state EvaluateDoublePress {
        [*] --> S2_Wait1: [time_since_init <= 500ms] PRESS
        S2_Wait1 --> S2_Wait2: [time_since_init <= 500ms] DEPRESS
    }
    EvaluateDoublePress --> if_state2 
    if_state2 --> DoublePress: [time_since_init <= 500ms] PRESS
    if_state2 --> SinglePress: [time_since_init > 500 ms] PRESS
    if_state2 --> Init: [time_since_init > 500 ms] DEPRESS

    SinglePress --> Init
    DoublePress --> Init
    LongPress --> Init
```

File: lib/atmega328/rotary-encoder.cpp (hunk 7 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp]: I still cannot figure this one out. :/
When you say "The point in time when the function is called cannot be assumed!". Do you mean the ISR should evaluate whether the button press is single, double, or long?

**Review by emanuelen5 — APPROVED** [review]

File: lib/atmega328/rotary-encoder.cpp (hunk 8 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp:86]: I see. Well, for sure it will be fun to create another state machine! 😄
I will resolve this comment but I already referenced it to an issue so we don´t forget that this needs to be done.

File: lib/atmega328/rotary-encoder.cpp (hunk 7 above)
Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp]: After some conversation, this can be resolved like this:
Disable interrupts when the evaluation happens.
* Recording SREG
* Clear (cli() ) 
* Re-establish SGER

Comment (nperozzi) [reply lib/atmega328/rotary-encoder.cpp]: I disabled interrupts while running this function.
I am resolving this comment but I guess that it will be fully resolved once the button library is implemented.
